### PR TITLE
Fix the definition of sigevent

### PR DESCRIPTION
**Pull request by asomers**
It was originally defined back before rust could represent C unions.  So
instead of defining the union field correctly, it simply defined that
union's most useful field.  Define it correctly now.

Include a backwards-compatibility mechanism: Rename sigevent's old
definition and hide it.  Implement Deref and DerefMut from sigevent to
the old definition, so consumers will still be able to use the old field
name.

### CODE REVIEW

**Comment by asomers** [comment]
@Amanieu what can we do about backwards-compat?  It should only be a problem for crates that uses the fairly obscure `sigev_notify_thread_id` field.  That includes Nix, which I can fix.  Can we use crater in "grep mode" just to find affected crates?

**Comment by JohnTitor** [comment]
>  Can we use crater in "grep mode" just to find affected crates?

AFAIK no, the cheapest way is `cargo check`.

**Comment by JohnTitor** [comment]
@Amanieu Semi-related, is there any discussion/progress about libc crate versioning? I'd like to resolve the root cause of such an issue someday...

**Comment by Amanieu** [comment]
No, there's not been any discussion on this. We could bring it up at the next libs team meeting, but I'm not really sure what needs to be done here.

**Comment by asomers** [comment]
Then here's what I suggest:
* Run Crater in check mode on all of libc's direct dependencies.
* If Nix is the only affected crate, then we can make the change, and publish a near synchronous patch release of Nix uses the newer libc crate.
* If there are a lots of crates that would be affected, then we change the PR to create a `sigevent2` structure and deprecate the original.  At some future date we rename `sigevent2` to `sigevent` and make `sigevent2` a deprecated type alias.

Does that sound reasonable?  Is craterbot already activated for this repo?

**Comment by Amanieu** [comment]
Crater is really designed to test builds of the Rust compiler & standard library, not external crates like libc. I'm sure it would be possible to adapt it, but I'm not familiar with crater and wouldn't know where to start.

As an alternative, it's possible to download the source code for all crates published on crates.io: https://twitter.com/m_ou_se/status/1433085053056262144

This would allow grepping for any uses of `sigev_notify_thread_id` to see if any crates other than nix would break from this change.

**Comment by asomers** [comment]
I used a command inspired by that Twitter post to download and search every crate version.  What I found was:
* Nix
* A few Nix forks, all since abandoned
* The autd-timer and autd3-timer crates, both abandoned and deleted from their upstream repo with the comments "start over" and "start over v2"
* The async-timer crate, which used this symbol up until v0.2.7, but then stopped.  And FWIW it never used that symbol correctly.
* A bunch of false positives

**Comment by asomers** [comment]
@Amanieu do you approve of the plan to fix libc and deal with the breakage by immediately publishing patch releases of Nix?

**Comment by Amanieu** [comment]
I'm concerned about the amount of breakage: there's a lot of crates that aren't using the latest major version of Nix. Do you think we could implement a backward-compatibility hack using `Deref`/`DerefMut` so that `.sigev_notify_thread_id` continues to work?

**Comment by asomers** [comment]
We would definitely patch more than just the current major version.  I think I should be able to do the top five or six.  But what about this DerefMut proposal.  How would that work?

**Comment by Amanieu** [comment]
If we implement `DerefMut` for `sigevent` such that it dereferences to an anonymous union with a `sigev_notify_thread_id`, existing users of `sigev_notify_thread_id` should still continue working. We also don't need to worry about accessing uninitialized data since reading out of a union is still unsafe.

**Comment by asomers** [comment]
That's a good idea, and it looks like it works!  I'll amend the PR.

**Comment by asomers** [comment]
@Amanieu the remaining CI failure is due to a limitation in the style checker.  It doesn't understand that the type declaration is part of an impl block.  I'm not sure how to fix that.  Do you have any ideas?

File: src/unix/bsd/freebsdlike/freebsd/mod.rs
Comment (Amanieu) [inline src/unix/bsd/freebsdlike/freebsd/mod.rs:1009]: I would prefer if this was a `union` that just contains a single field. This forces the access to the `sigev_notify_thread_id` field to be in an `unsafe` block, which is necessary since this may be an uninitialized field in the union.

**Comment by Amanieu** [comment]
Not sure about how to deal with the style checker, @JohnTitor any ideas?

**Comment by JohnTitor** [comment]
> Not sure about how to deal with the style checker, @JohnTitor any ideas?

Fixed in #2824

File: src/unix/bsd/freebsdlike/freebsd/mod.rs
Comment (asomers) [reply src/unix/bsd/freebsdlike/freebsd/mod.rs:1009]: But the whole point of this structure is for backwards-compatibility with older libc versions, and they didn't require `unsafe`.  Wouldn't it break backwards-compatibility then, if we put this field into a union?

Comment (Amanieu) [reply src/unix/bsd/freebsdlike/freebsd/mod.rs:1009]: It just needs to be backwards-compatible enough for existing users, which is basically just nix. Does nix access this field outside of an unsafe block?

Comment (asomers) [reply src/unix/bsd/freebsdlike/freebsd/mod.rs:1009]: Yes, existing Nix does not use `unsafe`.

Comment (asomers) [reply src/unix/bsd/freebsdlike/freebsd/mod.rs:1009]: Also, I don't think that it should be considered unsafe to access the union member when using the backwards-compatible definition.  Since it's defined as a struct with a bunch of padding fields, the whole structure should always be initialized, and it shouldn't be a problem to read from the "union" member.

**Comment by asomers** [comment]
Rebased and squashed.

File: src/unix/bsd/freebsdlike/freebsd/mod.rs
Comment (Amanieu) [reply src/unix/bsd/freebsdlike/freebsd/mod.rs:1009]: I think this is most likely fine, but I would like a second opinion from @JohnTitor.

Comment (JohnTitor) [reply src/unix/bsd/freebsdlike/freebsd/mod.rs:1009]: Hmm, I'm not sure if I've got 100% of the context here so don't take my thought as a strong opinion, but I think it doesn't have to be _perfectly correct_ while it's just for back-compat.

**Comment by asomers** [comment]
Fourth time's the charm :crossed_fingers:

**Comment by asomers** [comment]
Ok, I got the Android CI stuff running locally.  It turns out that Android requires a `skip_field_type` for `_sigev_un`.  I don't know why the other OSes don't.  I'm going to squash and merge.

File: src/unix/bsd/freebsdlike/freebsd/mod.rs
Comment (dtolnay) [inline src/unix/bsd/freebsdlike/freebsd/mod.rs:1236]: This Deref impl exposes UB to safe code on freebsd.

```rust
fn main() {
    let sigevent = libc::sigevent {
        sigev_notify: 0,
        sigev_signo: 0,
        sigev_value: libc::sigval {
            sival_ptr: 0 as *mut _,
        },
        _sigev_un: libc::__c_anonymous_sigev_un {
            _kevent_flags: 0,
        },
    };
    let _val = sigevent.sigev_notify_thread_id;
}
```

Tested in miri using https://github.com/rust-lang/miri/pull/2221:

```console
$ cargo miri run --target x86_64-unknown-freebsd

error: Undefined Behavior: type validation failed: encountered uninitialized bytes, but expected initialized bytes
  --> src/main.rs:12:16
   |
12 |     let _val = sigevent.sigev_notify_thread_id;
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered uninitialized bytes, but expected initialized bytes
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
           
   = note: inside `main` at src/main.rs:12:16
```

@Amanieu I thought this exact unsoundness issue came up in a recent T-libs-api meeting. Did we end up deciding that this is "okay enough" as a tradeoff? I didn't see it discussed explicitly here on the PR.

Screenshot from 2022-06-20 21-58-56

File: src/unix/bsd/freebsdlike/freebsd/mod.rs
Comment (dtolnay) [inline src/unix/bsd/freebsdlike/freebsd/mod.rs:1455]: All these unsafe blocks are reading and comparing potentially uninitialized data from the union.

```rust
fn main() {
    let sigevent = libc::sigevent {
        sigev_notify: libc::SIGEV_THREAD,
        sigev_signo: 0,
        sigev_value: libc::sigval {
            sival_ptr: 0 as *mut _,
        },
        _sigev_un: libc::__c_anonymous_sigev_un {
            _kevent_flags: 0,
        },
    };
    let _eq = sigevent == sigevent;
}
```

```console
error: Undefined Behavior: type validation failed: encountered uninitialized raw pointer
    --> /git/libc/src/unix/bsd/freebsdlike/freebsd/mod.rs:233:1
     |
233  | / s! {
234  | |     pub struct aiocb {
235  | |         pub aio_fildes: ::c_int,
236  | |         pub aio_offset: ::off_t,
...    |
1013 | |     }
1014 | | }
     | |_^ type validation failed: encountered uninitialized raw pointer
     |
     = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
     = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
             
     = note: inside `<libc::__c_anonymous_sigev_thread as std::cmp::PartialEq>::eq` at /git/libc/src/unix/bsd/freebsdlike/freebsd/mod.rs:876:9
     = note: inside `<libc::sigevent as std::cmp::PartialEq>::eq` at /git/libc/src/unix/bsd/freebsdlike/freebsd/mod.rs:1445:29
note: inside `main` at src/main.rs:12:15
    --> src/main.rs:12:15
     |
12   |     let _eq = sigevent == sigevent;
     |               ^^^^^^^^^^^^^^^^^^^^
     = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)
```

File: src/unix/bsd/freebsdlike/freebsd/mod.rs
Comment (dtolnay) [inline src/unix/bsd/freebsdlike/freebsd/mod.rs:1475]: Same here.

```rust
fn main() {
    let sigevent = libc::sigevent {
        sigev_notify: libc::SIGEV_THREAD,
        sigev_signo: 0,
        sigev_value: libc::sigval {
            sival_ptr: 0 as *mut _,
        },
        _sigev_un: libc::__c_anonymous_sigev_un {
            _kevent_flags: 0,
        },
    };
    println!("{:?}", sigevent);
}
```

File: src/unix/bsd/freebsdlike/freebsd/mod.rs
Comment (dtolnay) [inline src/unix/bsd/freebsdlike/freebsd/mod.rs:1492]: Same here.

I think either these impls need to be removed because `sigev_notify` is not trustworthy, or constructing `sigevent` with a user-controlled `sigev_notify` needs to be unsafe in some way, or constructing `sigevent` needs to be disallowed altogether.

**Review by dtolnay — CHANGES_REQUESTED** [review]

File: src/unix/bsd/freebsdlike/freebsd/mod.rs
Comment (asomers) [reply src/unix/bsd/freebsdlike/freebsd/mod.rs:1236]: But that field will never be uninitialized, because no program should ever use both `_sigev_un` and `sigev_notify_thread_id`.  The latter only exists to provide backwards-compatibility with the pre-union code.  So anybody accessing that field should already have initialized the entire structure.

File: src/unix/bsd/freebsdlike/freebsd/mod.rs
Comment (asomers) [reply src/unix/bsd/freebsdlike/freebsd/mod.rs:1455]: I agree.  I think we should remove most trait impls for any struct that contains a union (see #2816).  But how to do it without breaking existing code?

File: src/unix/bsd/freebsdlike/freebsd/mod.rs
Comment (dtolnay) [reply src/unix/bsd/freebsdlike/freebsd/mod.rs:1236]: I get that what you said is what programs _should_ do. If the libc crate ensures that programs do as you described, then that's great. If the **programmer** needs to ensure they do as you described or else they get memory unsafety (which is what is currently implemented in this PR) then we need to make them write `unsafe` somewhere.

Comment (asomers) [reply src/unix/bsd/freebsdlike/freebsd/mod.rs:1236]: The problem is that the old, non-union version of `sigevent` is already in the wild.  And since it isn't `unsafe`, there simply isn't any way for libc to fix the definition of `sigevent`, be backwards-compatible, *and* be `unsafe`.  The deref compromise suggested by @Amanieu is IMHO the best compromise.
FWIW there is exactly one currently-maintained crate on crates.io that uses the old `sigev_notify_thread_id` field, and I'm a maintainer of that crate, so I can fix any resulting bugs.  It's my intention to release a new version of Nix as soon as libc releases a new version with this PR.  However, the backwards-compatibility is still useful for users of older Nix versions.

Comment (dtolnay) [reply src/unix/bsd/freebsdlike/freebsd/mod.rs:1236]: > there simply isn't any way for libc to fix the definition of `sigevent`, be backwards-compatible, *and* be `unsafe`

Before this PR, libc::sigevent cannot be constructed in safe code because it has private padding fields. After this PR it can be constructed as shown above, which is a soundness bug. Preserving the property that it cannot be constructed in safe code is not a breaking change. The safety requirement on constructing a sigevent can be that you never expose it in a way that would mismatch the sigev_notify with the active variant of _sigev_un.

You can use #\[non_exhaustive\] if libc uses that these days, or just a zero sized private field with appropriate comment.

**Review by dtolnay — CHANGES_REQUESTED** [review]
`_private` looks good.

The 1.13.0 build is still going to fail like in https://github.com/rust-lang/libc/runs/6977334685 though. Please either get `cargo +1.13.0 build` working, or open a separate PR to remove the 1.13.0 job from CI because that is a ridiculous version to support.

**Comment by asomers** [comment]
@Amanieu @dtolnay I don't know how to satisfy both the style checker and Rust 1.13.0.  Older versions of Rust (tested on 1.14.0) seem to expand the macro incorrectly.  If I place the `#[cfg(libc_union)]` attribute directly over the struct, then recent rustc versions work but 1.14.0 seems to ignore it when expanding the s! macro.  Hard to be sure exactly, because cargo-expand does not work with Rust 1.14.0.  Placing `#[cfg(libc_union)]` before s! works, but that fails the style checker.

What do you want to do ?  Raise the MSRV to 1.19.0? Relax the style checker?  Or is there a version of cargo-expand that works with such ancient Rust versions, which might help me debug the macro?

**Comment by JohnTitor** [comment]
> open a separate PR to remove the 1.13.0 job from CI because that is a ridiculous version to support.

Why such aggressive language? Are you mocking the maintenance efforts that have been made?
Whatever the policy is, I'm hesitant to break it with a patch version while it's alive. And I don't think following it is "ridiculous".

> What do you want to do ? Raise the MSRV to 1.19.0? Relax the style checker? Or is there a version of cargo-expand that works with such ancient Rust versions, which might help me debug the macro?

I prefer to tweak the style checker, maybe adding support for "ignore this block" or "ignore this file" would be sufficient.

**Comment by asomers** [comment]
BTW, the current changes in this PR include the required style checker tweaks.  However, if https://github.com/rust-lang/libc/pull/2845 concludes , then I'll be able to revert the style checker changes and go back to preferred style.

**Comment by asomers** [comment]
@dtolnay since #2845 has stalled, would you consider merging this PR, despite its style changes?

**Comment by workingjubilee** [comment]
This PR has seen a lot of action, and it starts with a note of "...will fail to build with these changes". I skimmed it and am not sure where it's at right now, so I'm marking it as a breakage candidate since we're likely to start evaluating things we want to pull in for a 0.3 release real soon now, and I don't want this to get missed in the review for that, **even if it is not breaking anything**. However if you could make the PR's opening message match what the current status of this PR is, that would be greatly appreciated.

**Comment by asomers** [comment]
> This PR has seen a lot of action, and it starts with a note of "...will fail to build with these changes". I skimmed it and am not sure where it's at right now, so I'm marking it as a breakage candidate since we're likely to start evaluating things we want to pull in for a 0.3 release real soon now, and I don't want this to get missed in the review for that, **even if it is not breaking anything**. However if you could make the PR's opening message match what the current status of this PR is, that would be greatly appreciated.

Updated, @workingjubilee !  The current status of this PR is that, as far as I am concerned, it is complete.  However, there is a trilemma: we cannot have both backwards-compatibility AND rust 1.13.0 support AND an unmodified style checker.  The current state of the PR relaxes the style checker, but I would be ok with relaxes either or both of the other conditions, too.

**Comment by asomers** [comment]
@workingjubilee as explained in the current PR summary, this change should not cause any breakage.  Can we move forward with it?  And as far as the trilemma goes, which is your preference to drop?

* 100% Backwards-compatibility (though only one crate will be affected)
* Rust 1.13.0 support
* An unmodified style checker

**Comment by workingjubilee** [comment]
:thinking:

**Comment by workingjubilee** [comment]
Let me see if I can get a reduced version of #2813 through.

**Comment by asomers** [comment]
@workingjubilee do you need help with said reduced version?

**Comment by asomers** [comment]
@workingjubilee I rebased on main.  This version retains Rust 1.14.0 compatibility, but relaxes the style checker.

**Comment by asomers** [comment]
BTW, I don't think I ever explained the motivation for this change.  Without this change, it is impossible to use a multi-threaded Tokio runtime with POSIX AIO.  Only a single-threaded runtime can be used.  Switching to a multi-threaded runtime by using this libc branch improves my application's performance by 180% .  So the PR isn't just for correctness's sake; it's got real value too.

**Comment by SteveLauC** [comment]
If #3630 is a rebase of this PR, we can then close this PR? 🤔

**Comment by tgross35** [comment]
Considering the challenges this PR has had and how long it has been open, I think we are better off not worrying about 0.2 and just going forward with a fix for 1.0 via https://github.com/rust-lang/libc/pull/3630.

**Pull request closed without merging**
